benches/list.rs

```diff
@@ -0,0 +1,61 @@
+use criterion::{criterion_group, criterion_main, Bencher, BenchmarkId, Criterion};
+use ratatui::{
+    buffer::Buffer,
+    layout::Rect,
+    widgets::{List, ListItem, ListState, StatefulWidget, Widget},
+};
+
+/// Benchmark for rendering a list.
+/// It only benchmarks the render with a different amount of items.
+pub fn list(c: &mut Criterion) {
+    let mut group = c.benchmark_group("list");
+
+    for line_count in [64, 2048, 16384] {
+        let lines: Vec<ListItem> = (0..line_count)
+            .map(|_| ListItem::new(fakeit::words::sentence(10)))
+            .collect();
+
+        // Render default list
+        group.bench_with_input(
+            BenchmarkId::new("render", line_count),
+            &List::new(lines.clone()),
+            render,
+        );
+
+        // Render with an offset to the middle of the list and a selected item
+        group.bench_with_input(
+            BenchmarkId::new("render_scroll_half", line_count),
+            &List::new(lines.clone()).highlight_symbol(">>"),
+            |b, list| {
+                render_stateful(
+                    b,
+                    list,
+                    ListState::default()
+                        .with_offset(line_count / 2)
+                        .with_selected(Some(line_count / 2)),
+                )
+            },
+        );
+    }
+
+    group.finish();
+}
+
+/// render the list into a common size buffer
+fn render(bencher: &mut Bencher, list: &List) {
+    let mut buffer = Buffer::empty(Rect::new(0, 0, 200, 50));
+    bencher.iter(|| {
+        Widget::render(list.clone(), buffer.area, &mut buffer);
```

Comment:
I wonder if `list.clone()` call could be an expensive part of this?

Reply:
Yes it was, didn't find a solution for that originally. After more research I found [this post](https://stackoverflow.com/questions/69540182/benchmark-function-with-input-that-doesnt-implement-the-copy-trait) that dealt with this exact scenario.

To have a setup function, you have to use [`iter_batched`](https://docs.rs/criterion/0.5.1/criterion/struct.Bencher.html#method.iter_batched).
See [this doc](https://docs.rs/criterion/0.5.1/criterion/enum.BatchSize.html) for why I chose `LargeInput`. Short answer is
> `SmallInput` [...] is small enough that millions of values can be safely held in memory

and after testing, it uses less memory and runs about 10% faster.

Using `iter_batched` also runs about 20% faster, according to criterion output.

---

Here are the new performance graphs

![image](https://github.com/ratatui-org/ratatui/assets/36198422/4efbc7e8-535c-4521-ad09-19f6d5ccbc23)
![image](https://github.com/ratatui-org/ratatui/assets/36198422/cd8c6ee5-cd2a-4929-9763-6af98f8f1e61)

---

I tested this on other benches and all were improved, paragraph saw an improvement of 65%!
I'm going to submit a PR next to address that.

Reply:
Nice one! I thought of the paragraph issue when I saw this too, but wasn't sure how to solve it.